lib/fs/fspfs.c

```diff
@@ -0,0 +1,389 @@
+}
+
+static result_t fspfs_remove_empty_directory(void *data) {
+	struct inode *inode = (struct inode*)data;
+	result_t r;
+
+	return ifilesystem_delete_directory(inode->fs, inode->name);
+}
+
+static result_t fspfs_rename(void *data, const char *newpath) {
+	struct inode *inode = (struct inode*)data;
+	result_t r;
+	char newpath_buf[0x301];
+
+	if (sizeof(newpath_buf) < strlen(newpath))
+		return LIBTRANSISTOR_ERR_FS_NAME_TOO_LONG;
+	strncpy(newpath_buf, newpath, 0x301);
+
+	if (inode->is_dir)
+		// newpath must not exist, or be an existing empty directory.
+		r = ifilesystem_rename_directory(inode->fs, inode->name, newpath_buf);
+	else
+		// newpath will be replaced with oldpath. The behavior if newpath is a
+		// directory isn't really specified...
+		r = ifilesystem_rename_file(inode->fs, inode->name, newpath_buf);
+
+	// If this is a success, we need to fix this... Also, might want to have a
+	// way to invalidate the traversal.
+	if (r == RESULT_OK)
+		strncpy(inode->name, newpath, 0x301);
+
+	return r;
+}
+
+static result_t fspfs_open_as_file(void *data, int flags, int *fd) {
+	struct inode *inode = (struct inode*)data;
+	result_t r;
+	uint32_t ifs_flags;
+	uint64_t file_size;
+
+	if (inode->is_dir)
+		return LIBTRANSISTOR_ERR_FS_NOT_A_FILE;
+
+	if (flags & O_RDWR)
+		ifs_flags = 3;
+	else if (flags & O_WRONLY)
+		ifs_flags = 2;
+	else
+		ifs_flags = 1;
+	// In unix land, we're kinda sorta *always* in append mode.
+	ifs_flags |= 4;
+
+	struct ifs_file *f = malloc(sizeof(struct ifs_file));
+	if (f == NULL)
+		return LIBTRANSISTOR_ERR_OUT_OF_MEMORY;
+
+	if ((r = ifilesystem_open_file(inode->fs, &f->file, ifs_flags, inode->name)) != RESULT_OK)
+		goto fail;
+
+	if (flags & O_TRUNC) {
+		if ((r = ifile_set_size(f->file, 0)) != RESULT_OK) {
+			printf("Got an error: %x\n", r);
+			goto fail;
+		}
+	}
+
+	if (flags & O_APPEND) {
+		if ((r = ifile_get_size(f->file, &file_size)) != RESULT_OK) {
+			printf("Got an error: %x\n", r);
+			goto fail;
+		}
+		f->head = file_size;
+	} else {
+		f->head = 0;
+	}
+
+	*fd = fd_create_file(&fspfs_file_ops, f);
+	if (*fd < 0) {
+		// TODO: We should have an errno_to_result
+		r = LIBTRANSISTOR_ERR_UNSPECIFIED;
+		goto fail_ifs;
+	}
+
+	return RESULT_OK;
+
+fail_ifs:
+	ipc_close(f->file);
+fail:
+	free(f);
+	return r;
+}
+
+static result_t fspfs_open_as_dir(void *data, trn_dir_t *out) {
+	struct inode *inode = (struct inode*)data;
+	result_t r;
+
+	if (!inode->is_dir)
+		return LIBTRANSISTOR_ERR_FS_NOT_A_DIRECTORY;
+
+	idirectory_t *dir = malloc(sizeof(idirectory_t));
+	if (dir == NULL)
+		return LIBTRANSISTOR_ERR_OUT_OF_MEMORY;
+
+	if ((r = ifilesystem_open_directory(inode->fs, dir, 3, inode->name)) != RESULT_OK)
+		goto fail;
+
+	out->data = (void*)dir;
+	out->ops = &trn_fspfs_dir_ops;
+	return RESULT_OK;
+
+fail:
+	free(dir);
+	return r;
+}
+
+static result_t fspfs_release(void *data) {
+	struct inode *inode = (struct inode*)data;
+
+	ipc_close(inode->fs);
+	free(data);
+	return RESULT_OK;
+}
```

Comment:
this is the same `release` function you use for inodes that aren't the root? won't that close the filesystem object when you close an inode that's not the root of this filesystem?

Reply:
You're right. Maybe I should wrap the ifilesystem_t in a refcount ?